xds/src/main/java/io/grpc/xds/XdsNameResolver2.java

```diff
@@ -0,0 +1,378 @@
+              .setConfig(config)
+              .setCommittedCallback(new SelectionCompleted())
+              .build();
+    }
+
+    private boolean retainCluster(String cluster) {
+      AtomicInteger refCount = clusterRefs.get(cluster);
+      if (refCount == null) {
+        return false;
+      }
+      int count;
+      do {
+        count = refCount.get();
+        if (count == 0) {
+          return false;
+        }
+      } while (!refCount.compareAndSet(count, count + 1));
+      return true;
+    }
+
+    private void releaseCluster(String cluster) {
+      int count = clusterRefs.get(cluster).decrementAndGet();
+      if (count == 0) {
+        clusterRefs.remove(cluster);
+        syncContext.execute(new Runnable() {
+          @Override
+          public void run() {
+            updateResolutionResult();
+          }
+        });
+      }
+    }
+  }
+
+  private void updateResolutionResult() {
+    Map<String, ?> rawServiceConfig =
+        generateServiceConfigWithLoadBalancingConfig(clusterRefs.keySet());
+    if (logger.isLoggable(XdsLogLevel.INFO)) {
+      logger.log(
+          XdsLogLevel.INFO, "Generated service config:\n{0}", new Gson().toJson(rawServiceConfig));
+    }
+    ConfigOrError parsedServiceConfig = serviceConfigParser.parseServiceConfig(rawServiceConfig);
+    Attributes attrs =
+        Attributes.newBuilder()
+            .set(XdsAttributes.XDS_CLIENT_POOL, xdsClientPool)
+            .set(InternalConfigSelector.KEY, new ConfigSelector())
+            .build();
+    ResolutionResult result =
+        ResolutionResult.newBuilder()
+            .setAttributes(attrs)
+            .setServiceConfig(parsedServiceConfig)
+            .build();
+    listener.onResult(result);
+  }
+
+  @SuppressWarnings("ModifyCollectionInEnhancedForLoop")  // ok for concurrent map
+  private class ConfigWatcherImpl implements ConfigWatcher {
+    @Override
+    public void onConfigChanged(ConfigUpdate update) {
+      Set<String> clusters = new HashSet<>();
+      for (Route route : update.getRoutes()) {
+        RouteAction action = route.getRouteAction();
+        if (action.getCluster() != null) {
+          clusters.add(action.getCluster());
+        } else if (action.getWeightedCluster() != null) {
+          for (ClusterWeight weighedCluster : action.getWeightedCluster()) {
+            clusters.add(weighedCluster.getName());
+          }
+        }
+      }
+      boolean receivedNewCluster = false;
+      for (String newCluster : clusters) {
+        if (clusterRefs.putIfAbsent(newCluster, new AtomicInteger(1)) == null) {
+          receivedNewCluster = true;
+        }
+      }
+      // Update service config to include newly added clusters.
+      if (receivedNewCluster) {
+        updateResolutionResult();
+      }
+      // Make newly added clusters selectable by config selector.
+      routes = update.getRoutes();
+      // Drops reference for deleted clusters, update service config to remove deleted clusters
```

Comment:
This says _what_ it does but does not provide a hint as to _why_. The ordering here is very important.

"Make newly added clusters selectable by config selector." could maybe change to "Make newly added clusters selectable by config selector and deleted clusters no longer selectable."

Reply:
Good point. Put them together to reflect the ordering.